Rename NOTES*, README*, VERSION, HACKING, LICENSE to .md or .txt

DDvO opened the pull request:
This implements https://github.com/openssl/openssl/pull/12098#issuecomment-641355034.

For the files with names changed to `.md`, numerous small layout issues had to be solved because those files are now interpreted as MarkDown. It's likely that I missed some of them but at least all those files look reasonable now.

I also had to update various references to the renamed files - hope I caught all of them (I noticed and corrected some cases where similar references to other files renamed earlier had not been updated). Where it made most sense I converted them to proper (clickable) MarkDown-style references.

levitte commented on CHANGES.md:
```diff
@@ -3067,12 +3073,12 @@ OpenSSL 1.1.0
-   "OSSL_HANDSHAKE_STATE" instead of an int. SSL_set_state() has been removed
-   altogether. The previous handshake states defined in ssl.h and ssl3.h have
-   also been removed.
+   with the old code (see [ssl/statem/README.md](ssl/statem/README.md) for
```
You're thinking of [link references](https://spec.commonmark.org/0.29/#link-reference-definitions).  Don't go there.

DDvO replied:
> Isn't there a shortcut version of this, like `[foobar.md]` ? Here and below.

This would be rather ambiguous and error-prone,
and I did not find this in https://daringfireball.net/projects/markdown/syntax#link.

levitte commented on Configure:
```diff
@@ -282,7 +282,7 @@ foreach ( reverse sort( 'aes', 'aria', 'bf', 'camellia', 'cast', 'des', 'dh',
 
 collect_information(
-    collect_from_file(catfile($srcdir,'VERSION')),
+    collect_from_file(catfile($srcdir,'VERSION.txt')),
```
If you really must, I would call it `.dat` rather than `.txt`

levitte replied:
Reasoning: it contains version data, rather than free form text

levitte replied:
Also, there are two more spots to touch:

https://github.com/openssl/openssl/blob/d9c2fd51e2e278bc3f7793a104ff7b4879f6d63a/dev/release-aux/release-version-fn.sh#L39

https://github.com/openssl/openssl/blob/d9c2fd51e2e278bc3f7793a104ff7b4879f6d63a/dev/release-aux/release-version-fn.sh#L105

DDvO replied:
Yeah, that's better.

DDvO replied:
> Also, there are two more spots to touch:

Ah, thanks. And I found and updated a further reference to `VERSION` in `dev/release.s`.

levitte commented on INSTALL.md:
```diff
@@ -47,8 +47,8 @@ Prerequisites
- * Perl 5 with core modules (please read [NOTES.PERL](NOTES.PERL))
- * The Perl module Text::Template (please read [NOTES.PERL](NOTES.PERL))
+ * Perl 5 with core modules (please read [NOTES-Perl.md](NOTES-Perl.mdL))
+ * The Perl module `Text::Template` (please read [NOTES-PERL.ml](NOTES-Perl.md))
```
NOTES-PERL.ml => NOTES-Perl.md

DDvO replied:
Fixed

levitte commented on INSTALL.md:
```diff
@@ -47,8 +47,8 @@ Prerequisites
  * A make implementation
- * Perl 5 with core modules (please read [NOTES.PERL](NOTES.PERL))
- * The Perl module Text::Template (please read [NOTES.PERL](NOTES.PERL))
+ * Perl 5 with core modules (please read [NOTES-Perl.md](NOTES-Perl.mdL))
```
Spurious L at the end

DDvO replied:
Fixed

levitte commented on NOTES-Android.md:
```diff
@@ -16,23 +16,23 @@
- Do not pass --cross-compile-prefix (as you might be tempted), as it will
+ You have to use `./Configure` and name your target explicitly; there are
+ `android-arm`, `android-arm64`, `android-mips`, `android-mip64`, `android-x86`
+ and `android-x86_64` (*MIPS targets are no longer supported with NDK R20+).
```
Backticks around `*MIPS*`

DDvO replied:
Done

levitte commented on NOTES-DJGPP.md:
```diff
@@ -11,28 +11,28 @@
- requires that PERL and the PERL module Text::Template also be
- installed (see NOTES.PERL).
+ requires that PERL and the PERL module `Text::Template` also be
+ installed (see [NOTES-Perl.m](NOTES-Perl.md)).
```
.m => .md

DDvO replied:
Oh, so many such mistakes - fixed as well.

levitte commented on NOTES-Perl.md:
```diff
@@ -55,14 +55,14 @@
  are just a few exceptions:
 
- Test::More         We require the minimum version to be 0.96, which
+    Test::More      We require the minimum version to be 0.96, which
```
It's unfortunate that markdown doesn't have a recognised syntax for definition lists, that would be suitable here.

That being said, I think it would be preferable to make a proper list of this, so something like this:

``` md
 * Test::More

   We require the minimum version to be 0.96, which appeared in Perl 5.13.4,
   because that version was the first to have all the features we're using.
   This module is required for testing only!  If you don't plan on running the tests,
   you don't need to bother with this one.

 * Text::Template

   This module is not part of the core Perl modules.  As a matter of fact, the core
   Perl modules do not include any templating module to date.  This module is
   absolutely needed, configuration depends on it.
```

levitte replied:
I wasn't overly careful with the line length, please adapt appropriately

DDvO replied:
> It's unfortunate that markdown doesn't have a recognised syntax for definition lists, that would be suitable here.

Indeed.

> That being said, I think it would be preferable to make a proper list of this,

Done.

levitte commented on NOTES-Valgrind.md:
```diff
@@ -14,34 +14,33 @@ Requirements
-3. OpensSSL compiled
-   See: INSTALL
+   See http://valgrind.org/downloads/current.html
+3. OpensSSL compiled, sSee [INSTALL.md](INSTALL.md)
```
Spurious s

DDvO replied:
Oops

levitte commented on NOTES-Valgrind.md:
```diff
@@ -14,34 +14,33 @@ Requirements
 Test behavior can be modified by adjusting environment variables.
 
-EXE_SHELL
+    EXE_SHELL
```
This indentation is ugly, please don't.

DDvO replied:
> This indentation is ugly, please don't.

I did that to get code block formatting, but changed it to 
```
`EXE_SHELL`
```
also for similar cases below.

levitte replied:
I totally understand the intention, but we must keep in mind that people are reading these files with `more` or `less` as well, so there's a fine line to walk to get it as readable as possible, as straight text as well as with MD rendition.

DDvO commented on CHANGES.md:
```diff
@@ -136,6 +136,12 @@ OpenSSL 3.0
    *Richard Levitte*
 
+ * Project text documents not yet having a proper file name extension
+   (HACKING, LICENSE, NOTES*, README*, VERSION) have been renamed to *.md
```
> quote the glob patterns so that they are not seen as emphasis markers.

All right, done, also for several older entries,
although MD seems to cope with them well.

DDvO commented on CHANGES.md:
```diff
@@ -14638,7 +14646,7 @@ ndif
 
    - New object identifiers are inserted in objects.txt, following
-     the syntax given in objects.README.
+     the syntax given in [`crypto/objects/README.md`](crypto/objects/README.md).
```
Removed them.

DDvO commented on CONTRIBUTING.md:
```diff
@@ -64,7 +64,7 @@ guidelines:
  5. When at all possible, patches should include tests. These can
     either be added to an existing test, or completely new.  Please see
-    test/README for information on the test framework.
+    test/README.md for information on the test framework.
```
Yep, done.

DDvO commented on Configurations/README-design.md:
```diff
@@ -4,17 +4,17 @@ Design document for the unified scheme data
 throughout the source tree.  These files hold the minimum information
 needed to build end product files from diverse sources.  See the
-section on build.info files below.
+section on `build.info` files below.
```
> Can you linkify this? (I don't know, I'm asking)

It would not make sense here because it talks about `build.info` files in general, not about a specific one.

DDvO commented on Configurations/README-design.md:
```diff
@@ -61,27 +60,27 @@ dependencies.
 
-This is the build.info file in 'apps/', one may notice that all file
-paths mentioned are relative to the directory the build.info file is
+This is the `build.info` file in apps/`, one may notice that all file
```
Oops, yes.

DDvO commented on Configurations/README-design.md:
```diff
@@ -130,17 +129,17 @@ library 'libssl' is built from the source file 'ssl/tls.c'.
 be used when building anything that will be part of these modules.
-Also, both modules depend on the library 'libcrypto' to function
+Also, both modules depend on the library `libcrypto` to function
 properly.  ossltest is explicitly linked with the static variant of
```
Yes, also for further instances there.

DDvO commented:
> I got about 25% through this PR.

Thanks @richsalz and @levitte for pointing out various mistakes; 
I fixed all of them and some further minor issues I found in the course of doing that.

> Did you run `make md-nits` on your machine before creating this PR?

Oh, never heard of that. 
So I've installed `mdl` and I'm working through the findings now...

levitte commented:
> So I've installed mdl and I'm working through the findings now...

Please be careful what conclusions you draw.  `mdl` is over-zealous on quite a lot of points (to the point of actually being *wrong*, or at least overly unflexible), until told otherwise `util/markdownlint.rb`.

levitte commented on NOTES-Perl.md:
```diff
@@ -55,18 +55,22 @@
+  the first to have all the features we're using.
+  This module is required for testing only!
+  If you don't plan on running the tests,
+  you don't need to bother with this one.
```
I purposefully intendented this block of text to align with `Test::More`, so that it would be part of that list item.

DDvO replied:
I see - done this way.

levitte commented on NOTES-Perl.md:
```diff
@@ -55,18 +55,22 @@
+  As a matter of fact, the core Perl modules do not
+  include any templating module to date.
+  This module is absolutely needed,
+  configuration depends on it.
```
Likewise with this block

DDvO replied:
Yep

levitte commented on NOTES-Valgrind.md:
```diff
@@ -17,21 +17,21 @@ Requirements
 to specify a different executable environment.
 
-    EXE_SHELL="`/bin/pwd`/util/wrap.pl valgrind --error-exitcode=1 --leak-check=full -q"
+`EXE_SHELL="`/bin/pwd`/util/wrap.pl valgrind --error-exitcode=1 --leak-check=full -q"`
```
You might want to change `` `/bin/pwd` `` to `$(/bin/pwd)`, or double the outermost backticks.  See https://spec.commonmark.org/0.29/#code-spans

DDvO replied:
> You might want to change `/bin/pwd` to $(/bin/pwd), or double the outermost backticks.

Thx for the hint, but not needed.
Due to the 4 leading spaces this is correctly recognized as code block.

levitte replied:
So you went back to ugly?

DDvO replied:
> So you went back to ugly?

I put back the four spaces and the result looks good, at least to me, see
https://github.com/siemens/openssl/blob/rename-NOTES-etc/NOTES-Valgrind.md#running-tests

levitte replied:
And if you read it with `less`?

DDvO replied:
I also see no issue with the `less` output:
```
This variable is used to specify the shell used to execute OpenSSL test
programs. The default wrapper (`util/wrap.pl`) initializes the environment
to allow programs to find shared libraries. The variable can be modified
to specify a different executable environment.

    EXE_SHELL="`/bin/pwd`/util/wrap.pl valgrind --error-exitcode=1 \
        --leak-check=full -q"

```

DDvO commented:
> Please be careful what conclusions you draw. `mdl` is over-zealous on quite a lot of points (to the point of actually being _wrong_, or at least overly unflexible), until told otherwise in `util/markdownlint.rb`.

Good point. For this reason I've added three more rule exceptions,
in particular to allow bare URL. I see no point in re-writing, e.g.,

```
 directly at http://www.delorie.com/pub/djgpp. For help on which
``` 
to
```
 directly at [http://www.delorie.com/pub/djgpp](http://www.delorie.com/pub/djgpp).
 For help on which
```
as this just adds redundancy, wastes space, 
and often adds the need for extra line breaks.

DDvO commented:
I'm done with all changes due to the reviews, and `make md-nits` is satisfied.

levitte commented on NOTES-Android.md:
```diff
@@ -1,6 +1,5 @@
- NOTES FOR ANDROID PLATFORMS
- ===========================
+NOTES FOR ANDROID PLATFORMS
+===========================
```
This is one spot where I think mdl is over zealous

DDvO replied:
> This is one spot where I think mdl is over zealous

I though so too, but excluding the respective rule was for some reason not sufficient.
Well, and un-indenting the first two lines in a couple of `.md` files was no big deal.

levitte commented on NOTES-Valgrind.md:
```diff
@@ -14,34 +13,34 @@ Requirements
 
-   EXE_SHELL="`/bin/pwd`/util/wrap.pl valgrind --error-exitcode=1 --leak-check=full -q"
+    EXE_SHELL="`/bin/pwd`/util/wrap.pl valgrind --error-exitcode=1 \
+        --leak-check=full -q"
```
Oh, sorry for all the fuss with this one.  As an example, this is actually fine

richsalz commented:
I didn't mean link references.  Some markdown processors, when you would otherwise have ```[link-text](link-text)``` allow you to omit the parentheses part.

And while I'm here, just wanted to say that this is *amazing* work!

DDvO commented:
I've just rebased this, fixing conflicts with the just merged 12098.
Anything more to do on this PR?

DDvO commented:
@levitte, since you reviewed this and in my view all requested changes are done,
can you approve this now?

DDvO commented:
Travis failures are unrelated.

romen commented:
> > Please be careful what conclusions you draw. `mdl` is over-zealous on quite a lot of points (to the point of actually being _wrong_, or at least overly unflexible), until told otherwise in `util/markdownlint.rb`.
> 
> Good point. For this reason I've added three more rule exceptions,
> in particular to allow bare URL. I see no point in re-writing, e.g.,
> 
> ```
>  directly at http://www.delorie.com/pub/djgpp. For help on which
> ```
> 
> to
> 
> ```
>  directly at [http://www.delorie.com/pub/djgpp](http://www.delorie.com/pub/djgpp).
>  For help on which
> ```
> 
> as this just adds redundancy, wastes space,
> and often adds the need for extra line breaks.

The sane alternative to disabling the bare URL warning is to write urls within `<>`: `<https://www.example.com/>`

DDvO commented on Configurations/README.md:
```diff
@@ -237,31 +236,30 @@ In each table entry, the following keys are significant:
 [1] as part of the target configuration, one can have a key called
-    'inherit_from' that indicate what other configurations to inherit
-    data from.  These are resolved recursively.
+  `inherit_from` that indicate what other configurations to inherit
```
Fixed

DDvO commented on NOTES-Perl.md:
```diff
@@ -57,18 +57,22 @@
+    the first to have all the features we're using.
+    This module is required for testing only!
+    If you don't plan on running the tests,
+    you don't need to bother with this one.
```
> So... Are we implicitly recommending that it is fine to install openssl from sources on a platform in which tests cannot be run?

Good question.
@levitte, @mattcaswell?

DDvO replied:
> So... Are we implicitly recommending that it is fine to install openssl from sources on a platform in which tests cannot be run?

This question is still unanswered...

DDvO replied:
@romen, I just move your question into a separate issue: #12317
since it is actually independent of the re-structuring of the doc files here
in order to prevent this PR from being needlessly blocked.

DDvO commented:
> A few comments below.

Thanks @romen.

> I also have read about 1/3 of the changes so far, and this looks great!

Pleased to hear!

> I would mention here (instead than repeating the same comment line by line) that `[foo/bar.md]` is equivalent to `[foo/bar.md](foo/bar.md)` and in many places we could prefer the implicit notation for readability and maintainability.

This would be great, but sadly, as mentioned meanwhile in https://github.com/openssl/openssl/pull/12232#issuecomment-648623226, I found that this is not supported at least by Firefox and Chrome/Chromium.

> The other general note that I hope we will be able to address is that it seems `git` (well, the github ui at least, I haven't checked out this PR locally yet to examine more closely) seems to be losing track of some of the renames and instead treating them as o"old file was deleted, here is a new file". This is bad to track changes/blame in the documentation (which is often important in navigating how the project evolved and when features have been added, changed, deprecated, etc.).
> IIRC our resident git guru @mspncp recently faced similar problems and of course he found a solution to provide git with enough hints that it was able to track file history/renames correctly. @mspncp can you guide us with your magic once more?

I think that in this case `git` (and the GitHub pages) does the right thing already.
For instance, `git log -p` shows things like
```
diff --git a/NOTES.ANDROID b/NOTES-Android.md
similarity index 100%
rename from NOTES.ANDROID
rename to NOTES-Android.md
```

DDvO commented:
Can we please finalize this PR soon?

Due to the vast number of changes in many doc files this tends to quickly collect vast numbers of conflicts. 
I've just rebased to cover the conflicts than came up recently.

levitte commented on INSTALL.md:
```diff
@@ -58,13 +58,13 @@ To install OpenSSL, you will need:
- * [NOTES.PERL](NOTES.PERL) - some notes on Perl
+ * [NOTES-Unix.md](NOTES-Unix.md) - notes for Unix like systems
+ * [NOTES-VMS.md](NOTES-VMS.md) - notes related to OpenVMS
+ * [NOTES-Windows.txt](NOTES-Windows.txt) - notes related to the Windows platform
```
Why is this one `.txt`?

DDvO replied:
This was recently discussed in some related thread.
The point is to help poor Windows user souls where the File Explorer usually does not know how to open/display `.md` files and is not even smart enough to attempt interpreting them as plain text.

levitte commented on test/README.md:
```diff
@@ -1,130 +1,153 @@
-Test OpenSSL
-============
+How to add recipes
+==================
```
Ewwww, I'm really not very keen on mixing stuff geared for OpenSSL developers with stuff for OpenSSL users.  If I was a user, having to go through all the stuff for developers would feel quite jarring.

DDvO replied:
This was a merge glitch that I was about to correct.
Are you okay if I move this section to the end of `test/README.md`,
or which again was the file name you were proposing (can't find it at the moment)?

levitte replied:
I would prefer a separate file name for the developer stuff.  Sorry, I don't remember what file name I proposed

DDvO replied:
Fine, so I've moved to info on adding tests from `test/README.md` to a new file `test/README-dev.md` 
and referenced it  in `INSTALL.md`.

levitte commented on test/README.md:
```diff
@@ -128,3 +128,161 @@ To stochastically verify that the algorithm that produces uniformly distributed
 
     $ ./util/wrap.sh test/bntest -stochastic
+
+Guidelines for test developers
```
Suggestion: `test/README-developers.md`

levitte replied:
Ah, I saw that you did that.  Cool.

levitte commented:
While you're correct things, there's this at line 2934:

```
   configuration in one of the `Configurations/*.conf~ files (in
```

It should probably be:


```
   configuration in one of the `Configurations/*.conf` files (in
```

levitte commented on NOTES-VMS.md:
```diff
@@ -68,7 +63,7 @@
  as running the test suite.
 
- *The following is not available on Alpha*
+ `The following is not available on Alpha`
```
Uhmmmm, why?  Italics was intended

DDvO replied:
Oh, don't know why. Fixed.

DDvO replied:
Meanwhile I recall why I did a change at this point:
md-nits did not like it, reporting
```
./NOTES-VMS.md:66: MD036 Emphasis used instead of a header
```

So now I've resorted to
```
### The following is not available on Alpha
```

DDvO commented:
> While you're correct things, there's this at line 2934:
> 
> ```
>    configuration in one of the `Configurations/*.conf~ files (in
> ```
> 
> It should probably be:
> 
> ```
>    configuration in one of the `Configurations/*.conf` files (in
> ```

Oops, thanks, fixed.

Strange that `mdl` does not report such an imbalanced use of "`".

I just went though the rest of `CHANGES.md` and found many occurrences of "`xyz'" 

and changed them to "`xyz`".

levitte commented on Configurations/README.md:
```diff
@@ -461,37 +464,37 @@ conditions based on something in the passed variables, for example:
+template file in `Configurations/` named like the build file, with
+`.tmpl` appended, or in case of possible ambiguity, a combination of
+the second `build_scheme` list item and the `build_file` name.  For
+example, if `build_file` is set to `Makefile, the template could be
```
Missing `` ` ``

DDvO replied:
Fixed, along with a couple of more unbalanced "`" in block quotations above (where they are rather harmless).

levitte commented:
> Strange that `mdl` does not report such an imbalanced use of "`".

Not at all, you *are* allowed to have multiline inline code blocks

DDvO commented:
> > Strange that `mdl` does not report such an imbalanced use of "`".
> 
> Not at all, you _are_ allowed to have multiline inline code blocks

Yes, but I would expect that the number of backticks in a file should still be even.
There seems to be an implicit (redendering) rule that the effect of a backtick stretches at most to the end of its paragraph.

DDvO commented:
> Similarly, there's this line:
> 
> ```
>   * Make `openssl version' output lines consistent.
> ```
> 
> where the `'` should be converted to a `` ` ``

Done.

DDvO commented:
Rebased this mostly for kicking Travis after it had timed out on two runs last time.

DDvO commented:
I had to handle merge conflicts again.
Can we please finalize this soon?

t-j-h commented:
Hopefully @levitte can add his approval and we can get this in ... it is nice work!

DDvO commented:
> Hopefully @levitte can add his approval and we can get this in ... it is nice work!

Thanks @t-j-h!

@levitte (and also @richsalz and @romen) had already given various good comments, which I addressed,
and it appears to me he simply did not notice that or had no time to check on this.
In any case your approval will be sufficient.

t-j-h commented:
Understood. The approval you require is there. I'm just hoping that @levitte can use the 24 hour delay before merge to also indicate his approval as well given you addressed all of his comments and he did a detailed review.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Partly squashed, merged - thanks!

levitte commented:
Well, it still seems weird to me that INSTALL.md contains quite thorough details about all aspects of configuration, building and installation, but that the testing details are thrown off to the side, so to say...  I'm not sure I'll understand the purpose.  But again, i seem to be a minority, so I'll accept this outcome.
